Add gpuCI build scripts

charlesbluca opened the pull request:
<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
forum https://numba.discourse.group/.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities here, please click the arrow besides
   "Create Pull Request" and choose "Create Draft Pull Request".
   When it's ready for review, you can click the button "ready to review" near
   the end of the pull request
   (besides "This pull request is still a work in progress".)
   The maintainers will then be automatically notified to review it.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

This PR is here to test that gpuCI is running properly; once it is working properly, we can retarget to the upstream repo and point gpuCI at that repo instead.

cc @gmarkall

charlesbluca committed:
First pass at gpuCI build scripts

gmarkall commented on buildscripts/gpuci/axis.yaml:
```diff
@@ -0,0 +1,22 @@
+- "11.2"
+
+CUDA_TOOLKIT_VER:
+- "9.0"
```
Numba only supports 9.2 onwards, so this can be removed.

gmarkall commented on buildscripts/gpuci/build.sh:
```diff
@@ -0,0 +1,61 @@
+nvidia-smi
+
+gpuci_logger "Activate conda env"
+. /opt/conda/etc/profile.d/conda.sh
```
Maybe we can create a new environment for testing if conda installing a cuda toolkit causes problems with constraints

charlesbluca replied:
Good idea! Ideally this will grab the cached versions of the relevant packages (`llvmlite`, `numpy`, `scipy`, etc.) and we'll only have to download `cudatoolkit`

gmarkall replied:
Hopefully, yeah - if the cudatoolkit download becomes a problem, is that a point at which we'd think about making a custom docker image?

gmarkall reviewed (commented):
Couple of quick comments

charlesbluca committed:
Don't need to test CUDA 9.0

charlesbluca committed:
Make separate env for Numba testing

gmarkall commented on buildscripts/gpuci/build.sh:
```diff
@@ -0,0 +1,66 @@
+                  "llvmlite" \
+                  "numpy" \
+                  "scipy" \
+                  "jinja2" \
```
I think jinja2 is not needed - I think this is just something the documentation says but is a hangover from a very old version of Numba (not that this matters too much).

charlesbluca replied:
```suggestion

```

gmarkall commented on buildscripts/gpuci/build.sh:
```diff
@@ -0,0 +1,66 @@
+conda activate numba
+
+gpuci_logger "Install numba"
+python setup.py build_ext --inplace
```
I think just `setup.py develop` is sufficient as it calls `build_ext`.

```suggestion
```

gmarkall reviewed (commented):
Just a couple of small comments on the diff, but to my eye this looks good!

charlesbluca committed:
Apply suggestions from code review

Co-authored-by: Graham Markall <535640+gmarkall@users.noreply.github.com>

charlesbluca commented:
rerun tests

charlesbluca commented:
rerun tests

ajschmidt8 commented:
rerun tests

ajschmidt8 commented:
rerun tests

ajschmidt8 commented:
rerun tests

charlesbluca committed:
Fix environment creation

ajschmidt8 commented:
rerun tests

gmarkall commented on buildscripts/gpuci/build.sh:
```diff
@@ -0,0 +1,64 @@
+gpuci_mamba_retry create -n numba -y \
+                  "python=${PYTHON_VER}" \
+                  "cudatoolkit=${CUDA_TOOLKIT_VER}" \
+                  "llvmlite" \
```
We need the development package of llvmlite for testing Numba master - I think this should do the right thing:

```suggestion
                  "numba/label/dev::llvmlite" \
```

charlesbluca committed:
Use development version of llvmlite

Co-authored-by: Graham Markall <535640+gmarkall@users.noreply.github.com>

gmarkall committed:
Document gpuCI

gmarkall committed:
README: Remove macOS from CUDA support list.

gmarkall committed:
build.sh: Align a hash in a comment

No functional change, just to improve the aesthetics of the script.

gmarkall committed:
gpuCI config changes based on PR #7499 feedback

- Rename conda env to numba_ci so it's not "numba", which will be less
  confusing (e.g. reading paths).
- Use Anaconda compilers.
- Remove Python version check, which comes from numba -s also.
- Remove conda list, which comes from numba -s also

A couple of other changes not based on the feedback:

- Install psutil, suggested by numba -s
- Run in verbose mode, so if there's a crash in a test we can see which
  one.

gmarkall committed:
gpuCI build.sh: Remove dead hasArg code
